eks/main.tf

```diff
@@ -1,17 +1,43 @@
 locals {
   account_id = data.aws_caller_identity.current.account_id
 
-  subnets = var.enable_nat_gateway ? module.vpc.private_subnets : module.vpc.public_subnets
-
-  # Generate all node pools possible combinations of subnets and instance types
-  node_pool_combinations = [for idx, np in flatten([
-    for subnet in local.subnets : [
-      for itype in var.node_pools_config.instance_types : {
-        subnet_id     = subnet
-        instance_type = itype
-      }
+  zones           = { for k, v in data.aws_subnet.subnets : k => v.availability_zone }
+  subnets_by_zone = { for id, subnet in data.aws_subnet.subnets : subnet.availability_zone => id }
+  subnets         = var.enable_nat_gateway ? module.vpc.private_subnets : module.vpc.public_subnets
+
+  # Generate all node pools possible combinations of subnets and node pools
+  node_pool_combinations = flatten([
+    for np in var.node_pools : [
+      for zone in(np.zones != null ? np.zones : keys(local.subnets_by_zone)) : [
+        {
+          name          = np.name
+          subnet_id     = local.subnets_by_zone[zone]
+          instance_type = np.instance_type
+          labels        = np.labels
+          taints        = np.taints
+          ami_type      = np.ami_type
+          desired_size  = np.desired_size
+          max_size      = np.max_size
+          min_size      = np.min_size
+          disk_size     = np.disk_size
+        }
+      ]
     ]
-  ]) : np]
+  ])
+
+  node_pool_defaults = {
+    ami_type  = "AL2_x86_64"
+    disk_size = 20
+  }
+}
+
+data "aws_subnet" "subnets" {
+  for_each = toset(local.subnets)
+  id       = each.value
+}
+
+output "vpc" {
+  value = local.subnets_by_zone
 }
```

Comment:
Would this not output the AZ rather than a VPC?

```hcl
output "vpc" {
  value = local.subnets
}
```
This would be true if ngw was enabled, but it's not definitive as you could just end up with a list of subnets if it's not enabled if I'm reading your local right.

Reply:
This output was here just for debugging. But now you pointed it, would be worth it to add a couple of outputs